html/modules/custom/docstore/src/Controller/DocumentTypeController.php

```diff
@@ -39,292 +49,496 @@ class DocumentTypeController extends ControllerBase {
-    $node_type = $this->loadNodeType($type);
+  public function createDocumentType(Request $request) {
+    // Get provider.
+    $provider = $this->requireProvider();
 
-    $data = $this->buildJsonOutput($node_type);
+    // Parse JSON.
+    $params = $this->getRequestContent($request);
 
-    return $this->createJsonResponse($data, 200);
+    // Make sure endpoints are fresh.
+    // @todo necessary? We already refresh after creating the document type.
+    $this->rebuildEndpoints();
+
+    /** @var \Drupal\docstore\ManageFields $manager */
+    $manager = new ManageFields($provider, '', $this->entityFieldManager, $this->entityTypeManager, $this->database);
+
+    // Create document type.
+    $node_type = $manager->createDocumentType($params);
+
+    // Rebuild endpoints.
+    $this->rebuildEndpoints();
+
+    // Reset the list of accessible document types.
+    $this->rebuildAccessibleResourceTypes('node_type');
+
+    // Invalidate cache.
+    Cache::invalidateTags(['document_types']);
+
+    $data = [
+      'message' => 'Document type created',
+    ] + $this->buildDocumentTypeJsonOutput($node_type);
+
+    return $this->createJsonResponse($data, 201);
   }
 
   /**
    * Update document type.
+   *
+   * @param string $type
+   *   Document type.
+   * @param \Symfony\Component\HttpFoundation\Request $request
+   *   API request.
+   *
+   * @return \Symfony\Component\HttpFoundation\JsonResponse
+   *   JSON response.
+   *
+   * @todo consolidate logic with VocabularyController::updateVocabulary().
    */
   public function updateDocumentType($type, Request $request) {
+    // Parse JSON.
+    $params = $this->getRequestContent($request);
+
     /** @var \Drupal\node\Entity\NodeType $node_type */
     $node_type = $this->loadNodeType($type);
 
-    // Get provider.
+    /** @var \Drupal\user\UserInterface $provider */
     $provider = $this->requireProvider();
 
-    // Parse JSON.
-    $params = $this->getRequestContent($request);
-
     // Make sure endpoints are fresh.
+    // @todo necessary? We already refresh after updating the document type.
     $this->rebuildEndpoints();
 
-    $manager = new ManageFields($provider, '', $this->entityFieldManager);
-    $data = $this->buildJsonOutput($manager->updateDocumentType($node_type->id(), $params));
+    /** @var \Drupal\docstore\ManageFields $manager */
+    $manager = new ManageFields($provider, '', $this->entityFieldManager, $this->entityTypeManager, $this->database);
+
+    // Update the node type.
+    $node_type = $manager->updateDocumentType($node_type->id(), $params);
+
+    // Rebuild endpoints.
+    $this->rebuildEndpoints();
+
+    // Reset the list of accessible document types.
+    $this->rebuildAccessibleResourceTypes('node_type');
+
+    // Invalidate cache.
+    Cache::invalidateTags(['document_types']);
+
+    // Response data.
+    //
+    // @todo for other resources, we only return the "Resource updated" message
```

Comment:
Yes for consistency

Reply:
Thanks. Another outdated todo, I'll remove it.